src/content-tags-drawer/ContentTagsCollapsible.jsx

```diff
@@ -36,20 +106,179 @@ import ContentTagsTree from './ContentTagsTree';
+    // and reloads (on refocus). We only want to make requests to the backend when
+    // the user is updating the tags.
+    if (updatingTags) {
+      setUpdatingTags(false);
+      const tags = checkedTags.map(t => decodeURIComponent(t.split(',').slice(-1)));
+      mutation.mutate({ contentId, taxonomyId: id, tags });
+    }
+  }, [contentId, id, checkedTags]);
+
+  const mergeTrees = (tree1, tree2) => {
+    const mergedTree = { ...tree1 };
+
+    const sortKeysAlphabetically = (obj) => {
+      const sortedObj = {};
+      Object.keys(obj)
+        .sort()
+        .forEach((key) => {
+          sortedObj[key] = obj[key];
+          if (obj[key] && typeof obj[key] === 'object') {
+            sortedObj[key].children = sortKeysAlphabetically(obj[key].children);
+          }
+        });
+      return sortedObj;
+    };
+
+    const mergeRecursively = (destination, source) => {
+      Object.entries(source).forEach(([key, sourceValue]) => {
+        const destinationValue = destination[key];
+
+        if (destinationValue && sourceValue && typeof destinationValue === 'object' && typeof sourceValue === 'object') {
+          mergeRecursively(destinationValue, sourceValue);
+        } else {
+          // eslint-disable-next-line no-param-reassign
+          destination[key] = sourceValue;
+        }
+      });
+    };
+
+    mergeRecursively(mergedTree, tree2);
+    return sortKeysAlphabetically(mergedTree);
+  };
+
+  // This converts the contentTags prop to the tree structure mentioned above
+  React.useEffect(() => {
+    const resultTree = {};
+
+    contentTags.forEach(item => {
+      let currentLevel = resultTree;
+
+      item.lineage.forEach((key, index) => {
+        if (!currentLevel[key]) {
+          const isExplicit = index === item.lineage.length - 1;
+          currentLevel[key] = {
+            explicit: isExplicit,
+            children: {},
+          };
+
+          // Populating the SelectableBox with "selected" (explicit) tags
+          const value = item.lineage.map(l => encodeURIComponent(l)).join(',');
+          // eslint-disable-next-line no-unused-expressions
+          isExplicit ? add(value) : remove(value);
+        }
+
+        currentLevel = currentLevel[key].children;
+      });
+    });
+
+    setAppliedContentTags(resultTree);
+  }, [contentTags]);
+
+  // This is out source of truth that represents the current state of tags in
+  // this Taxonomy as a tree. Whenever either the `appliedContentTags` (i.e. tags passed in
+  // the prop from the backed) change, or when the `addedContentTags` (i.e. tags added by
+  // selecting/unselecting them in the dropdown) change, the tree is recomputed.
+  const tagsTree = React.useMemo(() => (
+    mergeTrees(appliedContentTags, addedContentTags)
+  ), [appliedContentTags, addedContentTags]);
+
+  // Add tag to the tree, and while traversing remove any selected ancestor tags
+  // as they should become implicit
+  const addTags = (tree, tagLineage, selectedTag) => {
```

Comment:
Any handler functions should be wrapped in [the `React.useCallback` hook](https://react.dev/reference/react/useCallback#usecallback)

Reply:
So this function is being used inside the handler, does it also need to be wrapped in the React.useCallback if the handler itself is wrapped?

Reply:
Oh, no then it's not necessary... but in that case I would personally just move this function outside of the component altogether to become a top-level function.

BTW the rule generally is if you're using a function as a prop value, it should be wrapped in a hook. The point of a hook like `useCallback` is that if it's called multiple times, it still returns the exact same function (same instance), and it only re-declares the function if the array of dependencies changes. Otherwise, every time your parent component is rendered, a new function instance is initialized and when you pass it as a prop, react considers it a changed prop (it's an identical function but a different instance, so it's not the same by reference), and then will have to re-render the child component. And as you saw, sometimes the variables captured within your function get out of sync with the components.

Reply:
Got it, the `addTags` function uses dispatchers from the `useCheckboxSetValues` so I left it there, but I moved the other functions outside the component. https://github.com/openedx/frontend-app-course-authoring/pull/704/commits/7c02136e16cfb9fbdacf907848387abd48ac7ff4